pulley/src/decode.rs

```diff
@@ -0,0 +1,675 @@
+        reg: u8,
+    },
+}
+
+impl core::fmt::Debug for DecodingError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        core::fmt::Display::fmt(self, f)
+    }
+}
+
+impl core::fmt::Display for DecodingError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Self::UnexpectedEof { position } => {
+                write!(f, "unexpected end-of-file at bytecode offset {position:#x}")
+            }
+            Self::InvalidOpcode { position, code } => {
+                write!(
+                    f,
+                    "found invalid opcode {code:#x} at bytecode offset {position:#x}"
+                )
+            }
+            Self::InvalidExtendedOpcode { position, code } => {
+                write!(
+                    f,
+                    "found invalid opcode {code:#x} at bytecode offset {position:#x}"
+                )
+            }
+            Self::InvalidReg { position, reg } => {
+                write!(
+                    f,
+                    "found invalid register {reg:#x} at bytecode offset {position:#x}"
+                )
+            }
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for DecodingError {}
+
+/// An abstraction over any kind of bytecode stream.
+///
+/// There are two primary implementations:
+///
+/// 1. `SafeBytecodeStream`: A thin wrapper around an index into a `&[u8]`. This
+///    implementation is 100% safe code.
+///
+/// 2. `UnsafeBytecodeStream`: A thin wrapper over a raw pointer. This
+///    implementation is wildly unsafe and will result in memory unsafety and
+///    other terrors when given invalid bytecode, or even valid bytecode
+///    encoding a program that itself does not preserve memory safety.
+pub trait BytecodeStream: Copy {
+    /// The type of error that this bytecode stream produces on invalid
+    /// operations.
+    type Error;
+
+    /// Create an "unexpected end-of-stream" error at the current position.
+    fn unexpected_eof(&self) -> Self::Error;
+
+    /// Create an "invalid opcode" error at the current position.
+    fn invalid_opcode(&self, code: u8) -> Self::Error;
+
+    /// Create an "invalid extended opcode" error at the current position.
+    fn invalid_extended_opcode(&self, code: u16) -> Self::Error;
+
+    /// Create an "invalid register" error at the current position.
+    fn invalid_reg(&self, reg: u8) -> Self::Error;
+
+    /// Advance the stream by the given number of bytes.
+    fn advance(&mut self, bytes: usize);
+
+    /// Read a single byte from the stream.
+    fn get1(&mut self) -> Result<u8, Self::Error>;
+
+    /// Read two bytes from the stream.
+    fn get2(&mut self) -> Result<[u8; 2], Self::Error>;
+
+    /// Read four bytes from the stream.
+    fn get4(&mut self) -> Result<[u8; 4], Self::Error>;
+
+    /// Read eight bytes from the stream.
+    fn get8(&mut self) -> Result<[u8; 8], Self::Error>;
```

Comment:
Reading over this again, WDYT about something like:

```rust
fn advance<const N: usize>(&mut self) -> Result<[u8; N], Self::Error>;
```

It feels a little weird below that in `SafeBytecodeStream` the `advance` method looks like it can cause panics but I think it probably never does in practice. By folding everything into one method though it might be reasonable to make this more obvious that the safe stream never panics?